File: applications/webapp/src/core/port/dto/response/ticket/TicketResponseDTO.ts
import { TicketStatus } from "@/core/domain/enumerate/TicketStatus";
import { TicketType } from "@/core/domain/enumerate/TicketType";
import { MemberResponseDTO } from "@/core/port/dto/response/member/MemberResponseDTO";

export type TicketResponseDTO = {
  id: string;
  createdAt: string;
  updatedAt: string;
  version: number;
  member: MemberResponseDTO;
  staff?: MemberResponseDTO;
  status: TicketStatus;
} & (
  | {
      type: TicketType.SUBMISSION_PRINT;
      properties: {
        submissionId: string;
        attachmentId: string;
      };
    }
  | {
      type: TicketType.TECHNICAL_SUPPORT;
      properties: {
        description: string;
      };
    }
  | {
      type: TicketType.NON_TECHNICAL_SUPPORT;
      properties: {
        description: string;
      };
    }
);

File: applications/backend/core/src/main/kotlin/com/forsetijudge/core/port/dto/response/ticket/TicketResponseDTO.kt
package com.forsetijudge.core.port.dto.response.ticket

import com.forsetijudge.core.domain.entity.Ticket
import com.forsetijudge.core.port.dto.response.member.MemberResponseBodyDTO
import com.forsetijudge.core.port.dto.response.member.toResponseBodyDTO
import java.io.Serializable
import java.time.OffsetDateTime
import java.util.UUID

data class TicketResponseBodyDTO(
    val id: UUID,
    val createdAt: OffsetDateTime,
    val updatedAt: OffsetDateTime,
    val version: Long,
    val member: MemberResponseBodyDTO,
    val staff: MemberResponseBodyDTO? = null,
    val type: Ticket.Type,
    val status: Ticket.Status,
    val properties: Map<String, Any>,
) : Serializable

fun Ticket<*>.toResponseBodyDTO() =
    TicketResponseBodyDTO(
        id = id,
        createdAt = createdAt,
        updatedAt = updatedAt,
        version = version,
        member = member.toResponseBodyDTO(),
        staff = staff?.toResponseBodyDTO(),
        type = type,
        status = status,
        properties = properties,
    )
